In [35]:
import heapq

class huffman():
    def __init__(self,probs):
        self.probs = probs
        pair =  self.encoding()
        self.codebook = pair[0]
        self.l = pair[1]
        
    def encoding(self):
        codebook = {}
        heap = self.probs
        heapq.heapify(heap)
        l = 0
        # concat nodes with the lowerst two robabilities
        while len(heap) > 1:
            p1, node1 = heapq.heappop(heap)
            p2, node2 = heapq.heappop(heap)
            for c in node1:
                codebook[c] = "0" + codebook.get(c, "")
            for c in node2:
                codebook[c] = "1" + codebook.get(c, "")
            l +=(p1+p2)
            heapq.heappush(heap, (p1 + p2, node1 + node2))
        return codebook,l
    def get_codebook(self):
        return self.codebook
    
    def get_avgLength(self):
        return self.l


In [26]:
# part a initialization
input_ = [(1/9, 'A'), (1/9, 'E'), (1/9, 'I'), (1/18, 'B'),(1/18,'C'),(1/18,"D")
             ,(1/12,"α"),(1/12,"β"),(1/12,"γ")
             ,(1/12,"δ"),(1/12,"ǫ"),(1/12,"φ")]
source = [k for (x,k) in input_]
prob_diction = dict([(k,x) for (x,k) in input_])
prob_diction,source

({'A': 0.1111111111111111,
  'E': 0.1111111111111111,
  'I': 0.1111111111111111,
  'B': 0.05555555555555555,
  'C': 0.05555555555555555,
  'D': 0.05555555555555555,
  'α': 0.08333333333333333,
  'β': 0.08333333333333333,
  'γ': 0.08333333333333333,
  'δ': 0.08333333333333333,
  'ǫ': 0.08333333333333333,
  'φ': 0.08333333333333333},
 ['A', 'E', 'I', 'B', 'C', 'D', 'α', 'β', 'γ', 'δ', 'ǫ', 'φ'])

In [78]:
# part a n = 1, initilization
n1 = []
for i in source:
    p = prob_diction[i]
    n1.append((p,i))
n1

[(0.1111111111111111, 'A'),
 (0.1111111111111111, 'E'),
 (0.1111111111111111, 'I'),
 (0.05555555555555555, 'B'),
 (0.05555555555555555, 'C'),
 (0.05555555555555555, 'D'),
 (0.08333333333333333, 'α'),
 (0.08333333333333333, 'β'),
 (0.08333333333333333, 'γ'),
 (0.08333333333333333, 'δ'),
 (0.08333333333333333, 'ǫ'),
 (0.08333333333333333, 'φ')]

In [79]:
# temporary save n1
props_entropy = n1

In [76]:
# part a, n =1, compute avg_length 
n1_huffman = huffman(n1)
n1_huffman.get_avgLength()

3.583333333333333

In [34]:
# part a, n =1, codebook
n1_huffman.get_codebook()

{'B': '0100',
 'C': '0101',
 'D': '1010',
 'ǫ': '1011',
 'α': '1100',
 'β': '1101',
 'γ': '1110',
 'δ': '1111',
 'φ': '000',
 'A': '001',
 'E': '011',
 'I': '100'}

In [55]:
# part a n = 2, initilization
n2 = []
for i in source:
    for j in source:
        p = prob_diction[i]*prob_diction[j]
        n2.append((p,i+j))
n2[:5]

[(0.012345679012345678, 'AA'),
 (0.012345679012345678, 'AE'),
 (0.012345679012345678, 'AI'),
 (0.006172839506172839, 'AB'),
 (0.006172839506172839, 'AC')]

In [70]:

# part a n = 2, compute average length
n2_huffman = huffman(n2)
n2_huffman.get_avgLength()/2

3.557098765432098

In [69]:

# part a n = 3, initialization 
n3 = []
for i in source:
    for j in source:
        for k in source:
            p = prob_diction[i]*prob_diction[j]*prob_diction[k]
            n3.append((p,i+j+k))
n3[:5]

[(0.0013717421124828531, 'AAA'),
 (0.0013717421124828531, 'AAE'),
 (0.0013717421124828531, 'AAI'),
 (0.0006858710562414266, 'AAB'),
 (0.0006858710562414266, 'AAC')]

In [71]:
# part a n = 3
n3_huffman = huffman(n3)
n3_huffman.get_avgLength()/3

3.5548553955189757

In [88]:
import math
entropy = 0
for i in props_entropy:
    d = i[0]*math.log(1/i[0],2)
    entropy+=d
entropy

3.5441104177484006


### Conclusion:
<font size="4">When the block size increases, the average length per source symbol will approach to the entropy,which is 3.5441104177484006 </font>

#### part b, n = 1, we have two sets,which are S1 = e, S2 = g
#### Given y, we know which set the source is from 
#### We need to compute two huffman encoding for S1 and S2 respectively



In [94]:
# probs for S1
probs_s1 = [(2/9,"A"),(2/9,"E"),(2/9,"I"),(1/9,"B"),(1/9,"C"),(1/9,"D")]

# save the probs for s1 temporary
entropy_probs_s1 = probs_s1.copy()

s1_huffman = huffman(probs_s1)

In [39]:
# S1 codebook
s1_huffman.get_codebook()

{'B': '000', 'C': '001', 'D': '110', 'A': '111', 'E': '01', 'I': '10'}

In [40]:
# S1 avg length per source symbol
s1_huffman.get_avgLength()

2.5555555555555554

In [96]:
# probs for S2
probs_s2 = [(1/6,"α"),(1/6,"β"),(1/6,"γ")
             ,(1/6,"δ"),(1/6,"ǫ"),(1/6,"φ")]

# save the probs for s2 temporary
entropy_probs_s2 = probs_s2.copy()

s2_huffman = huffman(probs_s2)

In [97]:
s2_huffman.get_codebook()

{'ǫ': '100', 'α': '101', 'β': '110', 'γ': '111', 'δ': '00', 'φ': '01'}

In [98]:
# S2 avg length per source symbol
s2_huffman.get_avgLength()

2.6666666666666665

In [99]:
# weighted avg length per source symbol, n=1
(2.5555555555555554+2.6666666666666665)/2

2.6111111111111107

## Codebook for n=1
##### 'φ','E': 01
#### 'δ', 'I': 00
#### 'γ', 'A': 111
#### 'β', 'D': 110
#### 'α': 101
#### 'ǫ': 100
#### 'B': 000
#### 'C': 001


### Part b, n =2, we have eg, ee, gg, ge four cases


In [45]:
prob_ee = (3/9+3/18)*(3/9+3/18)
prob_eg = (3/9+3/18)*(6/12)
prob_ge = (3/9+3/18)*(6/12)
prob_gg = (6/12)**2
# their prob are all 1/4
prob_ee,prob_eg,prob_ge,prob_gg

(0.25, 0.25, 0.25, 0.25)

In [46]:
probs_diction_e = {"A":2/9,"E":2/9,"I":2/9,"B":1/9,"C":1/9,"D":1/9}
probs_diction_g = {"α":1/6,"β":1/6,"γ":1/6,"δ":1/6,"ǫ":1/6,"φ":1/6}
e = ["A","E","I","B","C","D"]
g = ["α","β","γ","δ","ǫ","φ"]

In [50]:
# initialization of probalities of ee
probs_ee = []
for i in e:
    for j in e:
        p = probs_diction_e[i]*probs_diction_e[j]
        probs_ee.append((p,i+j))
probs_ee[:5]

[(0.04938271604938271, 'AA'),
 (0.04938271604938271, 'AE'),
 (0.04938271604938271, 'AI'),
 (0.024691358024691357, 'AB'),
 (0.024691358024691357, 'AC')]

In [52]:
#initialization of probalities of eg
probs_eg=[]
for i in e:
    for j in g:
        p = probs_diction_e[i]*probs_diction_g[j]
        probs_eg.append((p,i+j))
probs_eg[:5]

[(0.037037037037037035, 'Aα'),
 (0.037037037037037035, 'Aβ'),
 (0.037037037037037035, 'Aγ'),
 (0.037037037037037035, 'Aδ'),
 (0.037037037037037035, 'Aǫ')]

In [54]:
#initialization of probalities of ge
probs_ge=[]
for i in g:
    for j in e:
        p = probs_diction_g[i]*probs_diction_e[j]
        probs_ge.append((p,i+j))
probs_ge[:5]

[(0.037037037037037035, 'αA'),
 (0.037037037037037035, 'αE'),
 (0.037037037037037035, 'αI'),
 (0.018518518518518517, 'αB'),
 (0.018518518518518517, 'αC')]

In [63]:
#initialization of probalities of gg
probs_gg = []
for i in g:
    for j in g:
        p = probs_diction_g[i]*probs_diction_g[j]
        probs_gg.append((p,i+j))
probs_gg[:5]

[(0.027777777777777776, 'αα'),
 (0.027777777777777776, 'αβ'),
 (0.027777777777777776, 'αγ'),
 (0.027777777777777776, 'αδ'),
 (0.027777777777777776, 'αǫ')]

In [59]:
# Construct four huffman encoding for eg, ge,ee,gg
huffman_ee = huffman(probs_ee)
huffman_eg = huffman(probs_eg)
huffman_ge = huffman(probs_ge)
huffman_gg = huffman(probs_gg)

In [60]:
# compute Avg length per source symbol 
avg_length_partb_n2 = 1/4*(huffman_ee.get_avgLength()+huffman_eg.get_avgLength()+huffman_ge.get_avgLength()+huffman_gg.get_avgLength())/2
avg_length_partb_n2

2.575617283950617

### Part b, n =3, we have 'eee', 'eeg', 'ege', 'egg', 'gee', 'geg', 'gge', 'ggg' eight cases, prob for each of them is 1/8


In [66]:

probs_eee = []
probs_eeg = []
probs_ege = []
probs_egg = []
probs_gee = []
probs_geg = []
probs_gge = []
probs_ggg = []
# eee
for i in e:
    for j in e:
        for k in e:
            p = probs_diction_e[i]*probs_diction_e[j]*probs_diction_e[k]
            probs_eee.append((p,i+j+k))
# eeg
for i in e:
    for j in e:
        for k in g:
            p = probs_diction_e[i]*probs_diction_e[j]*probs_diction_g[k]
            probs_eeg.append((p,i+j+k))
# ege
for i in e:
    for j in g:
        for k in e:
            p = probs_diction_e[i]*probs_diction_g[j]*probs_diction_e[k]
            probs_ege.append((p,i+j+k))
# egg
for i in e:
    for j in g:
        for k in g:
            p = probs_diction_e[i]*probs_diction_g[j]*probs_diction_g[k]
            probs_egg.append((p,i+j+k))
# gee
for i in g:
    for j in e:
        for k in e:
            p = probs_diction_g[i]*probs_diction_e[j]*probs_diction_e[k]
            probs_gee.append((p,i+j+k))
# geg
for i in g:
    for j in e:
        for k in g:
            p = probs_diction_g[i]*probs_diction_e[j]*probs_diction_g[k]
            probs_geg.append((p,i+j+k))
# gge 
for i in g:
    for j in g:
        for k in e:
            p = probs_diction_g[i]*probs_diction_g[j]*probs_diction_e[k]
            probs_gge.append((p,i+j+k))
# ggg 
for i in g:
    for j in g:
        for k in g:
            p = probs_diction_g[i]*probs_diction_g[j]*probs_diction_g[k]
            probs_ggg.append((p,i+j+k))

In [68]:
# Construct huffman for eight cases:
huff_all = [huffman(probs_eee),huffman(probs_eeg),huffman(probs_ege),huffman(probs_egg),
            huffman(probs_gee),huffman(probs_geg),huffman(probs_gge),huffman(probs_ggg)]
avg_length_partb_n3 = sum([x.get_avgLength() for x in huff_all])*1/8*1/3
avg_length_partb_n3

2.5628715134887985

In [108]:
# conditional probalities H(X|Y)
entropy_probs_s2,entropy_probs_s1

([(0.16666666666666666, 'α'),
  (0.16666666666666666, 'β'),
  (0.16666666666666666, 'γ'),
  (0.16666666666666666, 'δ'),
  (0.16666666666666666, 'ǫ'),
  (0.16666666666666666, 'φ')],
 [(0.2222222222222222, 'A'),
  (0.2222222222222222, 'E'),
  (0.2222222222222222, 'I'),
  (0.1111111111111111, 'B'),
  (0.1111111111111111, 'C'),
  (0.1111111111111111, 'D')])

In [107]:
# compute the conditional entropy
s1_entropy,s2_entropy = 0,0
for a in entropy_probs_s2:
    d_s2 = a[0]*math.log(1/a[0],2)
    s2_entropy+=d_s2
for b in entropy_probs_s1:
    d_s1 = b[0]*math.log(1/b[0],2)
    s1_entropy+=d_s1
conditional_entropy = 1/2*s1_entropy+1/2*s2_entropy
conditional_entropy

2.544110417748401

### Conclusion:
 <font size="4">When the block size increases, the average length per source symbol will approach to the conditional entropy H(X<sub>0</sub>|Y) which is 2.544110417748401 </font>